server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamFailureStore.java

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.cluster.Diff;
+import org.elasticsearch.cluster.SimpleDiffable;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.core.Nullable;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+
+/**
+ * Holds the data stream failure store metadata that enable or disable the failure store of a data stream. Currently, it
+ * supports the following configurations:
+ * - enabled
+ */
+public record DataStreamFailureStore(boolean enabled) implements SimpleDiffable<DataStreamFailureStore>, ToXContentObject {
+
+    public static final ParseField ENABLED_FIELD = new ParseField("enabled");
+
+    public static final ConstructingObjectParser<DataStreamFailureStore, Void> PARSER = new ConstructingObjectParser<>(
+        "failure_store",
+        false,
+        (args, unused) -> new DataStreamFailureStore((Boolean) args[0])
+    );
+
+    static {
+        PARSER.declareBoolean(ConstructingObjectParser.constructorArg(), ENABLED_FIELD);
+    }
+
+    public DataStreamFailureStore() {
+        this(true);
+    }
+
+    public DataStreamFailureStore(@Nullable Boolean enabled) {
+        this(enabled == null || enabled);
+    }
```

Comment:
I think we may want to pass the nullability through? If we maintain it as the tri-state `Boolean` instead of defaulting it to "on", then we leave ourselves room in the future if we wanted to have a cluster-level "turn failure store on for (everything|`logs-*`)" setting.

Reply:
I thought that would expressed by not having the failure store configured to begin with, like we do with the lifecycle. For example:

```
{
   "options": {
   }
}
```

This would mean we have no opinion on the failure store and we follow whatever is considered best practice. Is that sufficient? Or do you think people can still choose to configure other failure store parameters?

Reply:
I was initially thinking the same thing as Mary. Passing the nullability through would only be necessary to support the following configuration:
```json
{
  "data_stream_options": {
     "failure_store": {
         "lifecycle": {}
     }
  }
}
```
That is, a user could configure the failure store with (future) options, but they don't want to explicitly enable or disable it -- i.e. they let another component or index template make that decision. For instance, a user could then say "if something enabled the failure store, I want its retention to be this". I'm inclined to think there's value in that, but it might be a bit overkill. What do you think?

Reply:
Not sure, I guess it depends how we will do the template composition. I was trying to remain inline with the lifecycle. But I am open to reconsider both of them if this adds value. I just do not think they should have a different behaviour. 

If I am not mistaken the idea behind defaulting to true was that if you are configuring an aspect of lifecycle probably you want it enabled. We thought it would be unlikely to set retention for 3 days if someone decides that this data stream should be managed, but if not "I do not care". Someone can still disable it later, but we did not consider the scenario where you want a specific retention but you do not care if it gets applied or not.

What do you think? Should we revisit the previous hypothesis or follow the same idea here too.

Reply:
No I think that hypothesis makes sense. Then I agree there's not much value in making the `enabled` field `Nullable`. While we're at it, is there a reason the constructor parameter _is_ `Nullable`? Wouldn't it make sense to have that a primitive `boolean` as well?

Reply:
Good point, I thought this was encapsulating the default behaviour in the constructor but I can add this in the XContent parser and if we need it for composition or something we can add it later.